src/resources/espnow.cc

```diff
@@ -0,0 +1,237 @@
+
+// Copyright (C) 2022 Toitware ApS.
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; version
+// 2.1 only.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// The license can be found in the file `LICENSE` in the top level
+// directory of this repository.
+
+#include "../top.h"
+
+// #define TOIT_FREERTOS 1
+#define CONFIG_TOIT_ENABLE_ESPNOW 1
+
+#if defined(TOIT_FREERTOS) && defined(CONFIG_TOIT_ENABLE_ESPNOW)
+
+#include <esp_wifi.h>
+#include <esp_event.h>
+#include <esp_netif.h>
+#include <esp_now.h>
+#include <esp_log.h>
+
+#include "../objects_inline.h"
+#include "../process.h"
+#include "../resource.h"
+#include "../resource_pool.h"
+#include "../vm.h"
+
+#include "../event_sources/system_esp32.h"
+
+#define ESPNOW_DATAGRAM_NUM 16
+#define ESPNOW_TX_WAIT_US   1000
+
+namespace toit {
+
+const int kInvalidESPNow = -1;
+
+// Only allow one instance of provisioning running.
+static  ResourcePool<int, kInvalidESPNow> espnow_pool(
+  0
+);
+
+static SemaphoreHandle_t tx_sem;
+static esp_now_send_status_t tx_status;
+static QueueHandle_t rx_queue;
+
+struct DataGram {
+  uint8_t mac[6];
+  int len;
+  uint8_t buffer[0];
+};
+
+class ESPNowResourceGroup : public ResourceGroup {
+ public:
+  TAG(ESPNowResourceGroup);
+
+  ESPNowResourceGroup(Process* process);
+  ~ESPNowResourceGroup();
+};
+
+ESPNowResourceGroup::ESPNowResourceGroup(Process* process)
+  : ResourceGroup(process) {
+  tx_sem = xSemaphoreCreateCounting(1, 0);
+  ASSERT(tx_sem);
+
+  rx_queue = xQueueCreate(ESPNOW_DATAGRAM_NUM, sizeof(void *));
+  ASSERT(rx_queue);
+}
+
+ESPNowResourceGroup::~ESPNowResourceGroup() {
+  tx_sem = NULL;
+}
+
+MODULE_IMPLEMENTATION(espnow, MODULE_ESPNOW)
+
+static void espnow_send_cb(const uint8_t *mac_addr, esp_now_send_status_t status)
+{
+  tx_status = status;
+  portBASE_TYPE ret = xSemaphoreGive(tx_sem);
+  if (ret != pdTRUE) {
+    // ESP_LOGE("ESPNow", "Failed to give TX semaphore");
+  }
+}
+
+static void espnow_recv_cb(const uint8_t *mac_addr, const uint8_t *data, int data_len)
+{
+  struct DataGram *datagram = (struct DataGram *)malloc(sizeof(struct DataGram) + data_len);
```

Comment:
I have written a short style guide because we have so many unwritten rules.
Most of them align with the Google C++ style guide.
One of them is that we don't allow C-style casts.
See https://github.com/toitlang/toit/tree/master/src#casts

Reply:
OK, I will read this document and modify the code.